clients/src/main/java/org/apache/kafka/common/telemetry/internals/MetricNamingConvention.java

```diff
@@ -0,0 +1,119 @@
+ * limitations under the License.
+ */
+package org.apache.kafka.common.telemetry.internals;
+
+import org.apache.kafka.common.MetricName;
+
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * This class encapsulates naming and mapping conventions defined as part of
+ * <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability#KIP714:Clientmetricsandobservability-Metricsnamingandformat">Metrics naming and format</a>
+ */
+public class MetricNamingConvention {
+
+    private static final String NAME_JOINER = ".";
+    private static final String TAG_JOINER = "_";
+
+    // remove metrics as it is redundant for telemetry metrics naming convention
+    private final static Pattern GROUP_PATTERN = Pattern.compile("\\.(metrics)");
+
+    public static MetricNamingStrategy<MetricName> getClientTelemetryMetricNamingStrategy(String domain) {
+        return new MetricNamingStrategy<MetricName>() {
+            @Override
+            public MetricKey metricKey(MetricName metricName) {
+                Objects.requireNonNull(metricName, "metric name cannot be null");
+                String group = metricName.group() == null ? "" : metricName.group();
+                String rawName = metricName.name() == null ? "" : metricName.name();
+
+                return new MetricKey(fullMetricName(domain, group, rawName),
+                    Collections.unmodifiableMap(cleanTags(metricName.tags())));
+            }
+
+            @Override
+            public MetricKey derivedMetricKey(MetricKey key, String derivedComponent) {
+                Objects.requireNonNull(derivedComponent, "derived component cannot be null");
+                return new MetricKey(key.getName() + NAME_JOINER + derivedComponent, key.tags());
+            }
+        };
+    }
+
+    /**
+     * Creates a metric name given the domain, group, and name. The new String follows the following
+     * conventions and rules:
+     *
+     * <ul>
+     *   <li>domain is expected to be a host-name like value, e.g. {@code org.apache.kafka}</li>
+     *   <li>group is cleaned of redundant words: "-metrics"</li>
+     *   <li>the group and metric name is dot separated</li>
+     *   <li>The name is created by joining the three components, e.g.:
+     *     {@code org.apache.kafka.producer.connection.creation.rate}</li>
+     * </ul>
+     */
+    private static String fullMetricName(String domain, String group, String name) {
+        return domain
+            + NAME_JOINER
+            + cleanGroup(group)
+            + NAME_JOINER
+            + cleanMetric(name);
+    }
+
+    /**
+     * This method maps a raw name to follow conventions and cleans up the result to be more legible:
+     * <ul>
+     *  <li> converts names to lower hyphen case conventions
```

Comment:
I don't see any conversion into lower hyphen in this method?

Reply:
Corrected the comment, my bad.